server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,464 @@
+            .uniqueIndex(ComponentDtoFunctions.toKey());
+    }
+  }
+
+  private static Map<String, Component> getReportFilesByKey(Component root) {
+    final ImmutableMap.Builder<String, Component> builder = ImmutableMap.builder();
+    new DepthTraversalTypeAwareCrawler(
+      new TypeAwareVisitorAdapter(CrawlerDepthLimit.FILE, POST_ORDER) {
+        @Override
+        public void visitFile(Component file) {
+          builder.put(file.getKey(), file);
+        }
+      }).visit(root);
+    return builder.build();
+  }
+
+  private Map<String, File> getDbFileSourcesByKey(Map<String, ComponentDtoWithSnapshotId> dbFilesByKey, Set<String> removedFileKeys) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ImmutableMap.Builder<String, File> builder = ImmutableMap.builder();
+      for (String fileKey : removedFileKeys) {
+        ComponentDtoWithSnapshotId componentDto = dbFilesByKey.get(fileKey);
+        FileSourceDto fileSourceDto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, componentDto.uuid());
+        if (fileSourceDto != null) {
+          builder.put(fileKey, new File(componentDto.path(), fileSourceDto.getSrcHash(), on('\n').splitToList(fileSourceDto.getLineHashes())));
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  private Map<String, File> getReportFileSourcesByKey(Map<String, Component> reportFilesByKey, Set<String> addedFileKeys) {
+    ImmutableMap.Builder<String, File> builder = ImmutableMap.builder();
+    for (String fileKey : addedFileKeys) {
+      // FIXME computation of sourceHash and lineHashes might be done multiple times for some files: here, in ComputeFileSourceData, in
+      // SourceHashRepository
+      Component component = reportFilesByKey.get(fileKey);
+      SourceLinesHashesComputer linesHashesComputer = new SourceLinesHashesComputer();
+      SourceHashComputer sourceHashComputer = new SourceHashComputer();
+      try (CloseableIterator<String> lineIterator = sourceLinesRepository.readLines(component)) {
+        while (lineIterator.hasNext()) {
+          String line = lineIterator.next();
+          linesHashesComputer.addLine(line);
+          sourceHashComputer.addLine(line, lineIterator.hasNext());
+        }
+      }
+      builder.put(fileKey, new File(component.getReportAttributes().getPath(), sourceHashComputer.getHash(), linesHashesComputer.getLineHashes()));
+    }
+    return builder.build();
+  }
+
+  private ScoreMatrix computeScoreMatrix(Map<String, File> dbFileSourcesByKey, Map<String, File> reportFileSourcesByKey) {
+    int[][] scoreMatrix = new int[dbFileSourcesByKey.size()][reportFileSourcesByKey.size()];
+    int maxScore = 0;
+
+    int dbFileIndex = 0;
+    for (Map.Entry<String, File> dbFileSourceAndKey : dbFileSourcesByKey.entrySet()) {
+      File fileInDb = dbFileSourceAndKey.getValue();
+      int reportFileIndex = 0;
+      for (Map.Entry<String, File> reportFileSourceAndKey : reportFileSourcesByKey.entrySet()) {
```

Comment:
no need to compute score if file suffixes are different. Or even better, no need to compute score if file languages are different (see `Language#getFileSuffixes()`)